### PR TITLE
Add _TZE200_viy9ihs7 (ONNDO Smart thermostat) to devices/tuya.ts

### PR DESCRIPTION
**Pull request by fume**
This is my first PR on this repo and my first ever device added to z2m 😄 

I bought this smart thermostat from aliexpress: https://it.aliexpress.com/item/1005005822896810.html?spm=a2g0o.order_list.order_list_main.17.47113696taaElP&gatewayAdapt=glo2ita

It wasn't supported so i added it successfully via an external converter.
There are some DPs which i wasn't able to manage, you'll find info in the comments.

HTH, regards

### CODE REVIEW

**Comment by fume** [comment]
missed the eslint task, i'm going to fix it. Sorry.

**Comment by Koenkk** [comment]
Thanks and congrats with your first contribution to the project!

File: src/devices/tuya.ts [hunk 1]
```diff
@@ -3788,18 +3788,13 @@ const definitions: Definition[] = [
...
-        ],
+        fingerprint: tuya.fingerprint('TS0601', ['_TZE200_viy9ihs7']),
         model: 'ZWT198',
         vendor: 'TuYa',
-        description: 'ONNDO Smart thermostat',
+        description: 'ONNDO smart thermostat',
         fromZigbee: [tuya.fz.datapoints],
         toZigbee: [tuya.tz.datapoints],
-        onEvent: tuya.onEventSetLocalTime,
+        onEvent: tuya.onEvent({timeStart: '2000'}),
```
Comment (fume) [inline src/devices/tuya.ts:3797]: May i ask you why you changed this? With the other variant (onEventSetTime if i recall) i had the problem that the thermostat was resetting the date/time that's why i used the onEventSetLocalTime

**Comment by fume** [comment]
Any suggestion on how to manage the dp109 which is the schedule? Should i parse the bytes array manually or is there an utility which i didn't notice?

**Comment by Koenkk** [comment]
With some luck you can re-use the parsers from other devices, e.g. https://github.com/Koenkk/zigbee-herdsman-converters/blob/32eebe703255824501ed6c1d20a0603c4ec24851/src/devices/tuya.ts#L2615

**Comment by habitoti** [comment]
Hi @fume,
thanks for creating this support. Just paired such a device, however:

- when trying to set/change the upper temp, it throws an error "No converter available for 'upper_temp'".
- when setting the current heating set point (set value is or should be 21), it fails with "Exception while calling fromZigbee converter: Value '2' is not allowed, expected one of [object Object],[object Object]}"
- local temperature calibration behaves a bit unreliable. Trying e.g. to use the stepper to reduce by .1 degrees suddenly makes the calibration value display jump to 25.5. It seems to have generally huge issues with negative calibration values.
- is the Deadzone temperature really just possible in .5 steps? That seems a pretty coarse precision, especially for floor heatings 

Regards, habitoti

**Comment by fume** [comment]
> Hi @fume, thanks for creating this support. Just paired such a device, however:
> 
> * when trying to set/change the upper temp, it throws an error "No converter available for 'upper_temp'".
> * when setting the current heating set point (set value is or should be 21), it fails with "Exception while calling fromZigbee converter: Value '2' is not allowed, expected one of [object Object],[object Object]}"
> * local temperature calibration behaves a bit unreliable. Trying e.g. to use the stepper to reduce by .1 degrees suddenly makes the calibration value display jump to 25.5. It seems to have generally huge issues with negative calibration values.
> * is the Deadzone temperature really just possible in .5 steps? That seems a pretty coarse precision, especially for floor heatings
> 
> Regards, habitoti

the converter is working fine for me, at least i don't have such exceptions. Are you sure this is the same exact device as mine?
Honestly i didn't play that much with the calibration and the deadzone, but the sensitivity should be respectively .1 and .5 as also detailed in the instruction manual.

**Comment by habitoti** [comment]
Yes, it‘s the ONNDO smart thermostate. This integration was automatically picked when I paired the device. Looks also identical to the picture in HA.

Not being able to change the setpoint is probably the most severe issue. As this is a thermostate device, it should be possible to manipulate it from automations or through schedules.

image

**Comment by fume** [comment]
> Yes, it‘s the ONNDO smart thermostate. This integration was automatically picked when I paired the device. Looks also identical to the picture in HA.
> 
> Not being able to change the setpoint is probably the most severe issue. As this is a thermostate device, it should be possible to manipulate it from automations or through schedules.
> 
> image

if you enable the debug logs, you should be able to see the payload and the DP before the error message. I'm curious and would like to compare it with mine. When i developed the converter i tested it ONLY on my thermostat. I hope that the same exact model has the same exact payload and DP.

**Comment by habitoti** [comment]
For some magical reason (or maybe after HA restart... :-) ) the setpoint error doesn't occur any longer right now.

Here is what happens when I set the local temperature calibration to "-0.1" by clicking the down-stepper once:

```
Debug 2023-11-06 17:28:23Received MQTT message on 'zigbee2mqtt/0x3425b4fffe4c8b96/set' with data '{"local_temperature_calibration":-0.1}'
Debug 2023-11-06 17:28:23Publishing 'set' 'local_temperature_calibration' to '0x3425b4fffe4c8b96'
Info 2023-11-06 17:28:23MQTT publish: topic 'zigbee2mqtt/0x3425b4fffe4c8b96', payload '{"child_lock":"UNLOCK","current_heating_setpoint":22,"deadzone_temperature":0.5,"frost_protection":"OFF","last_seen":"2023-11-06T17:28:17+01:00","linkquality":43,"local_temperature":21.6,"local_temperature_calibration":-0.1,"preset":"manual","running_state":"heat","system_mode":"heat","upper_temp":null}'
Debug 2023-11-06 17:28:28Received Zigbee message from '0x3425b4fffe4c8b96', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,255],"type":"Buffer"},"datatype":2,"dp":19}],"seq":19200}' from endpoint 1 with groupID 0
Info 2023-11-06 17:28:28MQTT publish: topic 'zigbee2mqtt/0x3425b4fffe4c8b96', payload '{"child_lock":"UNLOCK","current_heating_setpoint":22,"deadzone_temperature":0.5,"frost_protection":"OFF","last_seen":"2023-11-06T17:28:28+01:00","linkquality":43,"local_temperature":21.6,"local_temperature_calibration":25.5,"preset":"manual","running_state":"heat","system_mode":"heat","upper_temp":null}'
Debug 2023-11-06 17:28:28Received Zigbee message from '0x3425b4fffe4c8b96', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,215],"type":"Buffer"},"datatype":2,"dp":3}],"seq":19456}' from endpoint 1 with groupID 0
Info 2023-11-06 17:28:28MQTT publish: topic 'zigbee2mqtt/0x3425b4fffe4c8b96', payload '{"child_lock":"UNLOCK","current_heating_setpoint":22,"deadzone_temperature":0.5,"frost_protection":"OFF","last_seen":"2023-11-06T17:28:28+01:00","linkquality":43,"local_temperature":21.5,"local_temperature_calibration":25.5,"preset":"manual","running_state":"heat","system_mode":"heat","upper_temp":null}'
```

As you can see, it's in the end reported back as 25.5 (instead of -0.1).

For the upper temperature, Debug level doesn't show anything else than the missing converter error.

**Comment by fume** [comment]
> For some magical reason (or maybe after HA restart... :-) ) the setpoint error doesn't occur any longer right now.
> 
> Here is what happens when I set the local temperature calibration to "-0.1" by clicking the down-stepper once:
> 
> ```
> Debug 2023-11-06 17:28:23Received MQTT message on 'zigbee2mqtt/0x3425b4fffe4c8b96/set' with data '{"local_temperature_calibration":-0.1}'
> Debug 2023-11-06 17:28:23Publishing 'set' 'local_temperature_calibration' to '0x3425b4fffe4c8b96'
> Info 2023-11-06 17:28:23MQTT publish: topic 'zigbee2mqtt/0x3425b4fffe4c8b96', payload '{"child_lock":"UNLOCK","current_heating_setpoint":22,"deadzone_temperature":0.5,"frost_protection":"OFF","last_seen":"2023-11-06T17:28:17+01:00","linkquality":43,"local_temperature":21.6,"local_temperature_calibration":-0.1,"preset":"manual","running_state":"heat","system_mode":"heat","upper_temp":null}'
> Debug 2023-11-06 17:28:28Received Zigbee message from '0x3425b4fffe4c8b96', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,255],"type":"Buffer"},"datatype":2,"dp":19}],"seq":19200}' from endpoint 1 with groupID 0
> Info 2023-11-06 17:28:28MQTT publish: topic 'zigbee2mqtt/0x3425b4fffe4c8b96', payload '{"child_lock":"UNLOCK","current_heating_setpoint":22,"deadzone_temperature":0.5,"frost_protection":"OFF","last_seen":"2023-11-06T17:28:28+01:00","linkquality":43,"local_temperature":21.6,"local_temperature_calibration":25.5,"preset":"manual","running_state":"heat","system_mode":"heat","upper_temp":null}'
> Debug 2023-11-06 17:28:28Received Zigbee message from '0x3425b4fffe4c8b96', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,215],"type":"Buffer"},"datatype":2,"dp":3}],"seq":19456}' from endpoint 1 with groupID 0
> Info 2023-11-06 17:28:28MQTT publish: topic 'zigbee2mqtt/0x3425b4fffe4c8b96', payload '{"child_lock":"UNLOCK","current_heating_setpoint":22,"deadzone_temperature":0.5,"frost_protection":"OFF","last_seen":"2023-11-06T17:28:28+01:00","linkquality":43,"local_temperature":21.5,"local_temperature_calibration":25.5,"preset":"manual","running_state":"heat","system_mode":"heat","upper_temp":null}'
> ```
> 
> As you can see, it's in the end reported back as 25.5 (instead of -0.1).
> 
> For the upper temperature, Debug level doesn't show anything else than the missing converter error.

what happens if you do the same from the zigbee2mqtt UI?

**Comment by habitoti** [comment]
This was actually within Z2M Details tab of the entity…

**Comment by fume** [comment]
Sorry, i didn't get it. Give me some time to do some tests on my own device. Probably, in the meantime, the best option is to open another issue instead of keep the conversation on this PR, which is already merged anyway.

**Comment by marcelhoogantink** [comment]
Hello,
I have bought this device too, and meanwhile i managed to set and get almost all exposed features, including the work- and holiday-schedule.
I hope to share my external converter here soon for testing.
Please keep me informed if you move to a new issue.

**Comment by fume** [comment]
> Hello, I have bought this device too, and meanwhile i managed to set and get almost all exposed features, including the work- and holiday-schedule. I hope to share my external converter here soon for testing. Please keep me informed if you move to a new issue.

The schedule is something I completely skipped, so if you managed to make it work i guess you can open another PR integrating your converter with the actual one.

**Comment by gtothill** [comment]
Apologies if this is the wrong thread for this comment, Im happy to post elsewhere if necessary. I have eight of these devices (_TZE200_viy9ihs7) - thanks to everyone who worked on making them available to Zigbee2MQTT.
I just noticed something though - although the devices report the attribute System mode, this appears to be the wrong way round. When the thermostat is locally set to timed mode on the device the attribute is reported as ‘manual’ in the Z2MQTT GUI. When set to manual on the device - auto is shown in Z2MQTT. When auto is set on the device and over ridden Z2MQTT reports temporary_manual correctly.
I don’t know how to start to fix this but thought you guys would!

lastly - I note that none of these devices report their battery state correctly to Z2MQTT.

Thanks in advance for insights into these issues.

**Pull request merged**